docs/design/repair-with-hashes.md

```diff
@@ -0,0 +1,79 @@
+
+Failed hash, signature, or public key checks may indicate cheating. Therefore audit must implement this same piece hash check logic. Changes to audit and the reporting of cheaters are otherwise outside the scope of this document.  
+
+## Rationale
+
+Both download and Reed Solomon decoding create more load for a satellite than hashing, so hashing should be preferred over downloading even a single piece.
+
+Downloading for repair is significantly different enough from streaming as to warrant a new, custom implementation.
+
+Using only the minimum number of pieces means that Reed-Solomon does not act as a check during repair. Hence hashing is used instead. While an uplink could potentially send signed bogus data to a storage node, the storage node would not be penalized by these actions. This requires that Audit implements a similar piece hash check instead of relying solely on Reed-Solomon encoding.
+
+The size of all piece hashes downloaded should be roughly equal to a default maximum segment size : 64MiB.  It seems preferable to keep this in memory over dealing with persistance to disk.
+
+## Implementation
+
+1. Add an optional `PieceHash` and `OrderLimit` fields to the `PieceDownloadResponse` protocol buffer for returning hashes.
+
+```message PieceDownloadResponse {
+    // Chunk response for download request
+    message Chunk {
+        int64 offset = 1;
+        bytes data = 2;
+    }
+    Chunk chunk = 1;
+    optional PieceHash hash = 2;
+    optional OrderLimit limit = 3;
+}```
+
+2. Alter the storage node code to populate `PieceHash` and `OrderLimit` when the `OrderLimit` `Action` is `GET_REPAIR`.
+
+3. Write code which can validate a piece using a `PieceHash` and `OrderLimit` as described in design.
+
+4. Write code which ensures that enough of the pieces downloaded by satellite share the same `PiecePublicKey` as described in design.
+
+5. Implement custom repair download code which uses a Reed-Solomon minimum number of pieces, persisting the downloaded pieces only in memory. If statistics are available, prefer to download from faster storage nodes using the power of two choices.
+
+6. Wire in validation logic to repair loop.  If a piece fails these checks, download another.
+
+7. If any pieces failed any of the above checks, update the pointer in metainfo, removing those pieces.
+
+7. Reassemble the segment using Reed-Solomon decoding and continue with repair as previously implemented.
+
+## Open issues
+
+We may want to get rid of distinguishing between repair_gets and gets... should we always return them on a full piece download?
+
+64MiB is current default max segment size, it doesn't have to be equal or close to 64MiB.  Should we have an option to persist to disk?
```

Comment:
There is a check on the satellite on `CommitSegment` that limits the segment size requested by the uplink. Thus, I think it is safe to have an in-memory only implementation.